### PR TITLE
fix: User Management

### PR DESCRIPTION
**Pull request by Jonpro03**

### CODE REVIEW

File: api/src/serge/utils/security.py [hunk 1]
```diff
@@ -15,15 +16,12 @@
...
-
-
 def verify_password(plain_password, hashed_password):
-    return pwd_context.verify(plain_password, hashed_password)
+    return bcrypt.checkpw(plain_password.encode('utf-8'), hashed_password.encode('utf-8'))
 
 
 def get_password_hash(password):
-    return pwd_context.hash(password)
+    return bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt())
```
Comment (gaby) [inline api/src/serge/utils/security.py:24]: This will change the salt every time the function is call.

File: api/pyproject.toml [hunk 2]
```diff
@@ -39,12 +39,12 @@ pytest = "^8.3.2"
 hypercorn = {extras = ["trio"], version = "^0.17.3"}
 
 pyjwt = "^2.8.0"
-passlib = {extras = ["bcrypt"], version = "^1.7.4"}
 python-jose = {extras = ["cryptography"], version = "^3.3.0"}
 aiofiles = "^23.2.1"
 python-multipart = "^0.0.9"
 debugpy = "^1.8.1"
 sqlalchemy = "^2.0.30"
+bcrypt = "^4.2.0"
```
Comment (gaby) [inline api/pyproject.toml:47]: We should use the build-in hashlib in Python, it has `PBKDF2`.

**Review by gaby — CHANGES_REQUESTED** [review]
Why are we adding another Dockerfile, Script, and Configuration? There's too many files already.

File: web/src/routes/account/+page.ts [hunk 3]
```diff
@@ -10,13 +10,18 @@ interface User {
...
+    .then((response) => {
+      if (response.status == 401) {
+        window.location.href = "/";
+      }
+      return response.json();
+    })
+    .catch((error) => {
+      console.log(error);
+      window.location.href = "/";
+    });
```
[the remark this answers is not in the document]
Comment (gaby) [reply web/src/routes/account/+page.ts:25]: The log is enough

File: web/src/routes/chat/[id]/+page.ts [hunk 4]
```diff
@@ -33,8 +33,17 @@ interface Response {
...
+    .then((response) => {
+      if (response.status == 401) {
+        window.location.href = "/";
+      }
+      return response.json();
+    })
+    .catch((error) => {
+      console.log(error);
+      window.location.href = "/";
+    });
```
[the remark this answers is not in the document]
Comment (gaby) [reply web/src/routes/chat/[id]/+page.ts:46]: The log is enough

File: api/pyproject.toml (hunk 2 above)
Comment (Jonpro03) [reply api/pyproject.toml:47]: In the original impl, passlib was abstracting bcrypt, so all passwords are currently bcrypt hashed. Changing the underlying encryption will break existing user accounts, and there's no password reset functionality (yet) to recover those accounts.

Also depending on who you ask, bcrypt>pbkdf2 for its resistance to GPU brute-forcing... but between the three, scrypt is the best for modern systems.
Still, no ability to reset passwords, so I'm leery of changing the hashing algo and even bcrypt is overkill for a project of this scope

File: api/src/serge/utils/security.py (hunk 1 above)
Comment (Jonpro03) [reply api/src/serge/utils/security.py:24]: This is the intended behavior. If you're curious, [this is my favorite explanation](https://www.usenix.org/publications/loginonline/bcrypt-25-retrospective-password-security) of salt-in-password hashing.

File: api/pyproject.toml (hunk 2 above)
Comment (gaby) [reply api/pyproject.toml:47]: There hasnt been any releases, so it won't break anyones system.

File: api/src/serge/utils/security.py (hunk 1 above)
Comment (gaby) [reply api/src/serge/utils/security.py:24]: Awesome

File: api/pyproject.toml (hunk 2 above)
Comment (gaby) [reply api/pyproject.toml:47]: @Jonpro03 The build-n hashlib has scrypt, that works for me. 

https://docs.python.org/3/library/hashlib.html#hashlib.scrypt

**Comment by Jonpro03** [comment]
> Why are we adding another Dockerfile, Script, and Configuration? There's too many files already.

The debug dockerfile enables the python debugger

**Comment by gaby** [comment]
> > Why are we adding another Dockerfile, Script, and Configuration? There's too many files already.
> 
> The debug dockerfile enables the python debugger

Add it back to the .dev file, but can we make so it doesnt wait for something to attach to the debugger?

**Comment by Jonpro03** [comment]
> > > Why are we adding another Dockerfile, Script, and Configuration? There's too many files already.
> > 
> > 
> > The debug dockerfile enables the python debugger
> 
> Add it back to the .dev file, but can we make so it doesnt wait for something to attach to the debugger?

Yeah, no problem

File: api/src/serge/utils/security.py [hunk 5]
```diff
@@ -15,15 +18,20 @@
...
+    new_hashed_password = hashlib.scrypt(plain_password.encode("utf-8"), salt=salt, n=8192, r=8, p=1, dklen=64)
+    return new_hashed_password == stored_password
 
 
 def get_password_hash(password):
-    return pwd_context.hash(password)
+    salt = os.urandom(16)
+    hashed_password = hashlib.scrypt(password.encode("utf-8"), salt=salt, n=8192, r=8, p=1, dklen=64)
+    salt_and_hash = salt + hashed_password
+    return base64.b64encode(salt_and_hash).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (Jonpro03) [reply api/src/serge/utils/security.py:34]: Done

File: api/src/serge/utils/security.py [hunk 6]
```diff
@@ -15,15 +18,20 @@
...
-pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")
-
 
 def verify_password(plain_password, hashed_password):
-    return pwd_context.verify(plain_password, hashed_password)
+    salt_and_hash = base64.b64decode(hashed_password.encode("utf-8"))
+    salt = salt_and_hash[:16]
+    stored_password = salt_and_hash[16:]
+    new_hashed_password = hashlib.scrypt(plain_password.encode("utf-8"), salt=salt, n=8192, r=8, p=1, dklen=64)
+    return new_hashed_password == stored_password
```
[the remark this answers is not in the document]
Comment (Jonpro03) [reply api/src/serge/utils/security.py:27]: Done

**Review by gaby — APPROVED** [review]

**Pull request merged**